test/reshape.jl

```diff
@@ -507,6 +489,97 @@ end
     @test eltype(typeof(sdf2.value)) === Float64
 end
 
+@testset "additional unstack tests" begin
+    df = DataFrame(id=repeat(1:3, inner=3),
+                   id2=repeat(1:3, inner=3),
+                   var=repeat('a':'c', 3),
+                   val=1:9)
+    @test unstack(df, :id, :var, :val) == DataFrame(id=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+    @test unstack(df, [:id, :id2], :var, :val) == unstack(df, :var, :val) ==
+          DataFrame(id=1:3, id2=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+
+    # an exertice on current unstack invariants
+    Random.seed!(1234)
+    for i in 1:16
+        df = df[Random.shuffle(1:9), :]
+        @test unstack(df, :id, :var, :val)[sortperm(unique(df.id)), [1; 1 .+ sortperm(unique(df.var))]] ==
+              DataFrame(id=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+        @test unstack(df, [:id, :id2], :var, :val) == unstack(df, :var, :val)
+        @test unstack(df, :var, :val)[sortperm(unique(df.id)), [1:2; 2 .+ sortperm(unique(df.var))]] ==
+              DataFrame(id=1:3, id2=1:3, a=1:3:7, b=2:3:8, c=3:3:9)
+
+        df2 = copy(df)
+        df2.id = PooledArray(df.id)
+        df2.var = PooledArray(df.var)
+        @test unstack(df2, :id, :var, :val)[sortperm(df2.id.pool), [1; 1 .+ sortperm(df2.var.pool)]] ==
```

Comment:
in particular here we have an undocumented behaviour of `groupby`. I have pushed https://github.com/JuliaData/DataFrames.jl/pull/2481/commits/4014fb4552c676f036d309b07d3c074020ab1618 in #2481 to cover this case.